lttng-kallsyms.c

```diff
@@ -57,6 +59,103 @@ int lttng_enumerate_kernel_symbols(struct lttng_session *session)
 }
 EXPORT_SYMBOL_GPL(lttng_enumerate_kernel_symbols);
 
+#ifdef CONFIG_MODULES
+
+/*
+ * This function is taken from the linux kernel's module.c file
+ */
+static const char *kallsyms_symbol_name(struct mod_kallsyms *kallsyms, unsigned int symnum)
+{
+	return kallsyms->strtab + kallsyms->symtab[symnum].st_name;
+}
+
+static
+int lttng_kallsyms_module_coming(struct module *mod)
+{
+	int ret = 0, i;
+	struct mod_kallsyms *kallsyms;
+
+	/* Inspired by and partly taken from linux kernel's
+	 * module.c file, module_kallsyms_on_each_symbol function */
+	kallsyms = rcu_dereference(mod->kallsyms);
```

Comment:
Why do you use rcu_dereference here ? We are within a module coming notifier, before load_module invokes do_init_module which does rcu_assign_pointer(mod->kallsyms, &mod->core_kallsyms);

So I don't see any point is doing a rcu dereference here, since we are serialized with respect to module loading.

Reply:
Because of this comment in module.c, function module_kallsyms_on_each_symbol (line 4153)

/* We hold module_mutex: no need for rcu_dereference_sched */
struct mod_kallsyms *kallsyms = mod->kallsyms;

Since we don't hold the module_mutex here, I thought we need the rcu_dereference call. But if you are sure we don't need it, I'll remove it.

Reply:
Because the module coming notifier, called from module loading while holding an initial reference to the module (only released within do_init_module after the coming notifier), we are guaranteed that no concurrent changes occur to this structure (they are not freed by module unloading because we still hold the initial reference).

The comment you refer to applies to accessing this structure from other execution contexts which are not part of the module notifiers.